Update for vector search index in 5.13

nilsceberg opened the pull request:
* drop beta tag for feature, and its create and query procedures
* keep beta tag on set property procedure
* deprecate old set procedure in favor of new procedure with different name and signature
* noted parameter validation issue (fixed in 5.13)
* noted tx state issue and workaround (fixed in 5.13)

nilsceberg committed:
Note additional known and fixed issues in the vector index

neo-technology-commit-status-publisher commented:
Thanks for the documentation updates.

The preview documentation has now been torn down - reopening this PR will republish it.
<!-- Sticky Pull Request Comment -->

parnmatt committed:
Drop beta tags for vector search feature

Vector search index and its create and query procedures are out of beta
The set vector procedures are still in beta

parnmatt committed:
Add new vector setting procedure

This new procedure superscedes the old procedure, which is now
deprecated.

renetapopova commented on modules/ROOT/pages/indexes-for-vector-search.adoc:
```diff
@@ -64,9 +63,13 @@ Returns the requested number of approximate nearest neighbor nodes and their sim
 | Set vector property.
 | {link-procedures-reference}#procedure_db_create_setVectorProperty[`db.create.setVectorProperty`]
-| Update a given node property with the given vector in a more space-efficient way than directly using xref:clauses/set.adoc#set-set-a-property[`SET`].
+| label:beta[] label:deprecated[] Please use {link-procedures-reference}#procedure_db_create_setNodeVectorProperty[`db.create.setNodeVectorProperty`]
```
We try to avoid using "please" in the docs.
```suggestion
| label:beta[] label:deprecated[] It is replaced by {link-procedures-reference}#procedure_db_create_setNodeVectorProperty[`db.create.setNodeVectorProperty`]
```

parnmatt committed:
Update modules/ROOT/pages/indexes-for-vector-search.adoc

Co-authored-by: Reneta Popova <reneta.popova@neo4j.com>

parnmatt committed:
Remove "please" later on

renetapopova commented on modules/ROOT/pages/indexes-for-vector-search.adoc:
```diff
@@ -267,31 +270,43 @@ Removed 1 index.
 ----
 
-The example above matches one node and updates its embedding, but it's also possible to use a list parameter containing several MATCH criterias and embeddings, to update multiple nodes in an `UNWIND` clause.
+The example above matches one node and updates its embedding, but it's also possible to use a list parameter containing several `MATCH` criteria and embeddings, to update multiple nodes in an `UNWIND` clause.
```
```suggestion
Furthermore, you can also use a list parameter containing several `MATCH` criteria and embeddings to update multiple nodes in an `UNWIND` clause.
```

renetapopova replied:
Could it be possible to provide an example?

parnmatt replied:
I believe @michael-simons added that section in #745

From my understanding of what he wrote, it's just normal cypher usage that a independent of the procedures.

My personal cypher knowledge is quite minimal, so I don't quite get _exactly_ what Michael was getting at here. Perhaps he has some input can could provide an example?

parnmatt replied:
It seems @michael-simons is on leave until Monday 16th.
I don't know what you want to do about this @renetapopova …perhaps this is ok for now, and further clarification could be added in a later PR?

renetapopova replied:
I agree, we can add an example later.

renetapopova commented on modules/ROOT/pages/indexes-for-vector-search.adoc:
```diff
@@ -64,9 +63,13 @@ Returns the requested number of approximate nearest neighbor nodes and their sim
 
+| Set vector property.
+| {link-procedures-reference}#procedure_db_create_setNodeVectorProperty[`db.create.setNodeVectorProperty`]
+| label:beta[] Update a given node property with the given vector in a more space-efficient way than directly using xref:clauses/set.adoc#set-set-a-property[`SET`].
```
```suggestion
| label:beta[] label:new[Introduced in 5.13] Update a given node property with the given vector in a more space-efficient way than directly using xref:clauses/set.adoc#set-set-a-property[`SET`]. Replaces  {link-procedures-reference}#procedure_db_create_setVectorProperty[`db.create.setVectorProperty`].
```

renetapopova commented on modules/ROOT/pages/indexes-for-vector-search.adoc:
```diff
@@ -391,8 +406,8 @@ For example, you cannot have one xref:indexes-vector-similarity-euclidean[Euclid
 
-The vector search index is a beta feature.
-The below table lists the known issues in its current implementation.
+The vector search index is was a beta feature until Neo4j 5.13.
```
```suggestion
As of Neo4j 5.13, the vector search index is no longer a beta feature.
```

renetapopova commented on modules/ROOT/pages/indexes-for-vector-search.adoc:
```diff
@@ -391,8 +406,8 @@ For example, you cannot have one xref:indexes-vector-similarity-euclidean[Euclid
-The vector search index is a beta feature.
-The below table lists the known issues in its current implementation.
+The vector search index is was a beta feature until Neo4j 5.13.
+The below table lists the known issues in its implementations.
```
```suggestion
The following table lists the known issues and the version in which they were fixed:
```

renetapopova commented on modules/ROOT/pages/indexes-for-vector-search.adoc:
```diff
@@ -10,11 +10,10 @@
 = Vector search index
 
-_Vector search indexes were released as a public beta in Neo4j 5.11._
+_Vector search indexes were released as a public beta in Neo4j 5.11, and general availability in Neo4j 5.13._
```
```suggestion
_Vector search indexes were released as a public beta in Neo4j 5.11 and general availability in Neo4j 5.13._
```

renetapopova commented on modules/ROOT/pages/indexes-for-vector-search.adoc:
```diff
@@ -267,31 +270,43 @@ Removed 1 index.
 They are represented as properties on nodes with type `LIST<FLOAT>`.
-The recommended way of setting a vector property is using the link:{link-procedures-reference}#procedure_db_create_setVectorProperty[`db.create.setVectorProperty`] procedure.
-This procedure validates the input and sets the property as an array of {ieee-754} single precision values.
+The recommended way of setting a vector property is using the {link-procedures-reference}#procedure_db_create_setNodeVectorProperty[`db.create.setNodeVectorProperty`] procedure.
```
```suggestion
As of Neo4j 5.13, you can set a vector property using the {link-procedures-reference}#procedure_db_create_setNodeVectorProperty[`db.create.setNodeVectorProperty`] procedure.
```

renetapopova commented on modules/ROOT/pages/indexes-for-vector-search.adoc:
```diff
@@ -267,31 +270,43 @@ Removed 1 index.
-The recommended way of setting a vector property is using the link:{link-procedures-reference}#procedure_db_create_setVectorProperty[`db.create.setVectorProperty`] procedure.
-This procedure validates the input and sets the property as an array of {ieee-754} single precision values.
+The recommended way of setting a vector property is using the {link-procedures-reference}#procedure_db_create_setNodeVectorProperty[`db.create.setNodeVectorProperty`] procedure.
+This _beta_ procedure validates the input and sets the property as an array of {ieee-754} single precision values.
```
```suggestion
It validates the input and sets the property as an array of {ieee-754} single precision values.
This beta procedure replaces {link-procedures-reference}#procedure_db_create_setVectorProperty[`db.create.setVectorProperty`].
```

renetapopova commented on modules/ROOT/pages/indexes-for-vector-search.adoc:
```diff
@@ -267,31 +270,43 @@ Removed 1 index.
 == Set a vector property on a node
 
 Valid vectors for use in the index must have components finitely representable in {ieee-754} _single_ precision.
 They are represented as properties on nodes with type `LIST<FLOAT>`.
```
```suggestion
They are represented as properties on nodes with the type `LIST<FLOAT>`.
```

renetapopova commented on modules/ROOT/pages/indexes-for-vector-search.adoc:
```diff
@@ -267,31 +270,43 @@ Removed 1 index.
 ----
 
-Usually you want to define your embeddings as Cypher parameters and call `db.create.setVectorProperty` as in the following example:
+[NOTE]
```
I would suggest removing this note.

parnmatt replied:
Is it clear enough without it?

renetapopova replied:
Same here, I added it as part of the text. See line 279.

renetapopova commented on modules/ROOT/pages/indexes-for-vector-search.adoc:
```diff
@@ -267,31 +270,43 @@ Removed 1 index.
+As of Neo4j 5.13 {link-procedures-reference}#procedure_db_create_setVectorProperty[`db.create.setVectorProperty`] is _deprecated_ and has been replaced by {link-procedures-reference}#procedure_db_create_setNodeVectorProperty[`db.create.setNodeVectorProperty`].
+====
+
+Usually you want to define your embeddings as Cypher parameters and call `db.create.setNodeVectorProperty` as in the following example:
```
```suggestion
The following example shows how to define embeddings as Cypher parameters by matching a node and setting its vector properties using `db.create.setNodeVectorProperty`:
```

renetapopova commented on modules/ROOT/pages/indexes-for-vector-search.adoc:
```diff
@@ -267,31 +270,43 @@ Removed 1 index.
+The example above matches one node and updates its embedding, but it's also possible to use a list parameter containing several `MATCH` criteria and embeddings, to update multiple nodes in an `UNWIND` clause.
 This is ideal for creating and setting new vector properties in the graph.
 
 You can also set a vector property on a node using the xref:clauses/set.adoc#set-set-a-property[`SET`] command:
```
```suggestion
You can also set a vector property on a node using the xref:clauses/set.adoc#set-set-a-property[`SET`] command as in the following example:
```

renetapopova commented on modules/ROOT/pages/indexes-for-vector-search.adoc:
```diff
@@ -301,12 +316,12 @@ You can also set a vector property on a node using the xref:clauses/set.adoc#set
+RETURN node
 ----
 
 However, Cypher stores the provided `LIST<FLOAT>` as a primitive array of {ieee-754} _double_ precision values.
```
```suggestion
However, Cypher stores the provided `LIST<FLOAT>` as a primitive array of {ieee-754} _double_ precision values. 
This takes up almost twice as much space compared to the alternative method, where you use the `db.create.setNodeVectorProperty` procedure.
As a result, using `SET` for a vector index is not recommended. 

To reduce the storage space, you can reset the existing properties using `db.create.setNodeVectorProperty`. 
However, this comes with the cost of an increase in the transaction log size until they are rotated away.
```

renetapopova commented on modules/ROOT/pages/indexes-for-vector-search.adoc:
```diff
@@ -301,12 +316,12 @@ You can also set a vector property on a node using the xref:clauses/set.adoc#set
 ----
 
 However, Cypher stores the provided `LIST<FLOAT>` as a primitive array of {ieee-754} _double_ precision values.
 As a result, it takes up approximately twice as much space and it's not recommended for properties that are used in a vector index.
```
```suggestion
```

renetapopova commented on modules/ROOT/pages/indexes-for-vector-search.adoc:
```diff
@@ -301,12 +316,12 @@ You can also set a vector property on a node using the xref:clauses/set.adoc#set
 However, Cypher stores the provided `LIST<FLOAT>` as a primitive array of {ieee-754} _double_ precision values.
 As a result, it takes up approximately twice as much space and it's not recommended for properties that are used in a vector index.
-Existing properties can be re-set using `db.create.setVectorProperty` to minimize store size, but there is a cost in the transaction log size until they are rotated away.
+Existing properties can be re-set using `db.create.setNodeVectorProperty` to minimize store size, but there is a cost in the transaction log size until they are rotated away.
```
```suggestion
```

parnmatt replied:
This might be important to still note in some way, shape or form, as people who are using the procedure would be concerned with space; and then may be quite surprised to see that (at least for a while) the total disk spaced used increases by quite a bit.

parnmatt replied:
If you have an alternative suggestion or rough idea which may be a better approach to express this, I'd be grateful, as I agree it doesn't sit right in its current state.

renetapopova replied:
I added it as part of the text and removed it here. But if you think it's not visible enough, I can take another look.

parnmatt replied:
Sorry @renetapopova I completely missed that. The diff didn't make it clear to me; however, I can see it clear as day when looking at the rendered page!

renetapopova commented on modules/ROOT/pages/indexes-for-vector-search.adoc:
```diff
@@ -391,8 +406,8 @@ For example, you cannot have one xref:indexes-vector-similarity-euclidean[Euclid
+The below table lists the known issues in its implementations.
 
 [%header,cols="5a,d"]
 |===
```
I would suggest that we start with the latest version on top.

renetapopova reviewed (changes requested):
I added some editorial suggestions.

parnmatt committed:
Apply suggestions from code review

Co-authored-by: Reneta Popova <reneta.popova@neo4j.com>

parnmatt committed:
Move known issues solved in 5.13 to the top of the table

parnmatt committed:
Update modules/ROOT/pages/indexes-for-vector-search.adoc

Co-authored-by: Reneta Popova <reneta.popova@neo4j.com>

parnmatt committed:
Remove note on replaced procedure

renetapopova reviewed (approved):
I think it looks good now. Thanks, @nilsceberg and @parnmatt.
